mlflow/johnsnowlabs.py

```diff
@@ -0,0 +1,874 @@
+        dst_dir,
+        spark_model,
+        mlflow_model,
+        sample_input,
+        conda_env,
+        code_paths,
+        signature=None,
+        input_example=None,
+        pip_requirements=None,
+        extra_pip_requirements=None,
+        remote_model_path=None,
+):
+    """
+    Saves model metadata into the passed-in directory.
+    If mlflowdbfs is not used, the persisted metadata assumes that a model can be
+    loaded from a relative path to the metadata file (currently hard-coded to "sparkml").
+    If mlflowdbfs is used, remote_model_path should be provided, and the model needs to
+    be loaded from the remote_model_path.
+    """
+    import pyspark
+
+    if sample_input is not None:
+        mleap.add_to_model(
+            mlflow_model=mlflow_model,
+            path=dst_dir,
+            spark_model=spark_model,
+            sample_input=sample_input,
+        )
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, dst_dir)
+
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, dst_dir)
+
+    # add the johnsnowlabs flavor
+    # TODO dynamic pyspark version # pyspark.__version__p
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        pyspark_version=johnsnowlabs.settings.raw_version_pyspark,
+        model_data=_JOHNSNOWLABS_MODEL_PATH_SUB,
+        code=code_dir_subpath,
+    )
+
+    # add the pyfunc flavor
+    pyfunc.add_to_model(
+        mlflow_model,
+        loader_module="mlflow.johnsnowlabs",
+        data=_JOHNSNOWLABS_MODEL_PATH_SUB,
+        conda_env=_CONDA_ENV_FILE_NAME,
+        python_env=_PYTHON_ENV_FILE_NAME,
+        code=code_dir_subpath,
+    )
+    mlflow_model.save(os.path.join(dst_dir, MLMODEL_FILE_NAME))
+
+    if conda_env is None:
+        default_reqs = get_default_pip_requirements() if pip_requirements is None else None
+        conda_env, pip_requirements, pip_constraints = _process_pip_requirements(
+            default_reqs,
+            pip_requirements,
+            extra_pip_requirements,
+        )
+    else:
+        conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)
+
+    with open(os.path.join(dst_dir, _CONDA_ENV_FILE_NAME), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    # Save `constraints.txt` if necessary
+    if pip_constraints:
+        write_to(os.path.join(dst_dir, _CONSTRAINTS_FILE_NAME), "\n".join(pip_constraints))
+    # TODO REMOVE THIS HACK PR IS MERGED
+    pip_requirements = [r for r in pip_requirements if not r.startswith('mlflow==2.2')]
+    write_to(os.path.join(dst_dir, _REQUIREMENTS_FILE_NAME), "\n".join(pip_requirements))
+
+    _PythonEnv.current().to_yaml(os.path.join(dst_dir, _PYTHON_ENV_FILE_NAME))
+
+    save_jars_and_lic(dst_dir)
+
+
+def save_jars_and_lic(dst_dir):
+    deps_data_path = os.path.join(dst_dir, _JOHNSNOWLABS_MODEL_PATH_SUB, "jars.jsl")
```

Comment:
Instead of mixing the os.path.join and pathlib.Path paradigms, can we just stick to pathlib for path manipulations, joining, directory creation, and file writing?

Reply:
thanks, replaced all occurences of `os.pah.join` with `Pathlib.joinpath`